Improve `from` search prefix error handling

aescling opened the pull request:
addresses #36

note https://github.com/mastodon/mastodon/pull/17963; i expect they will ask for adjustments there; i further expect to make impactful design changes. i am still opening this PR as we are currently running this on GlitchCat anyway, with no reported problems

aescling committed:
Make `from` search prefix more robust (addresses mastodon/mastodon#17941)

* Improve robustness for account string validation

Using unsupported prefixes now reports a 422; searching for posts from an
account the instance is not aware of reports a 404. TODO: The UI for this
on the front end is abysmal.

Searching `from:username@domain` now succeeds when `domain` is the local
domain; searching `from:@username(@domain)?` now works as expected.

* Satisfy upstream rubcocp

* Unbreak upstream tests

single-right-quote commented:
updated for upstream rubocop and upstream tests

single-right-quote commented:
we will not be incorporating upstream&#x2019;s requested changes

marrus-sh commented on app/models/concerns/account_finder_concern.rb:
```diff
@@ -25,8 +25,21 @@ def find_local(username)
+    end
   end
 
+  ACCOUNT_STRING_RE = /^@?(?<username>#{Account::USERNAME_RE})(?:@(?<domain>[[:word:]\.\-]+))?$/i
```
`[:word:]` includes characters outside of the ASCII range, I’m pretty sure. If Mastodon allows IDNs elsewhere then that’s fine, but I’m asking because I’m not sure. If it doesn’t, use `\w`, which is just ASCII.

marrus-sh replied:
I’d also like to move this RegExp somewhere else eventually, but I don‘t think that blocks this MR.

single-right-quote replied:
i literally copypasted `MENTION_RE` (pre-#39) and added anchors to the sides

marrus-sh replied:
okay cool lol; it’d be nice if we could just have one regexp instead of two but uhhh i know grouping can be a pain with that sometimes

anyway that wouldn’t block me from merging this

single-right-quote replied:
yeah i would have defined it in terms of MENTION_RE if i knew a way to

single-right-quote replied:
> i literally copypasted `MENTION_RE` (pre-#39) and added anchors to the sides and named matching

oh actually i made a mistake: actual `MENTION_RE` uses `[[:word:]\.\-]+[[:word]]+' for the `domain` part. so currently a `from:@a@a` would run a search, whereas if we actually duplicated mention processing behavior (which i indeded and would still want), we wouldn&#x2019;t match on `@a@a` and return a (still presently unhelpful) error

marrus-sh commented:
It’d be nice to have tests for at least the new methods, so that if something upstream breaks this we would have some manner of indication. I wouldn’t rate it as an especially high priority though.

single-right-quote commented on app/controllers/api/v2/search_controller.rb:
```diff
@@ -11,6 +11,16 @@ class Api::V2::SearchController < Api::BaseController
+  rescue Mastodon::SyntaxError
+    unprocessable_entity
+  # user searched for posts from an account the instance is not aware of
+  rescue Mastodon::NotFound
```
i have no strong opinion but it Feels Wrong

we *could* steal from upstream and rely on `Account.find_remote!` throwing `ActiveRecord::NotFound`, which would remove the need for our new custom error

aescling committed:
Make account string validation consistent with mention processing

We previously matched on one-character domains and domains ending with
`[\.-]`, allowing `from:@a@a` and `from:@a@a-` searches to cause an
account lookup. This commit will raise a syntax error in both cases, as
MENTION_RE would never match them.

aescling committed:
Refactor `from` prefix error handling.

Incorporates changes suggested in #37. In doing so, adopts an error
handling style more consistent with the existing codebase (for which I
must thank @ClearlyClaire).

Removes new code no longer in use.

single-right-quote commented:
i really should add tests but my smoke tests (running these commits briefly on GlitchCat2 and testing that searches work, and and that searchs 422 on one-character domain names or domain names ending not in `:word:`) raised no problems

(we really need to set up a test instance)

marrus-sh reviewed (approved):
LGTM. Feel free to merge if you don’t want to bother with tests yet; maybe at least run rubocop though if you haven’t since we don’t have CI.

single-right-quote commented:
rubocop reports no (new) offenses; i&#x2019;m gonna merge and then immediately file an issue to write tests, since we really ought to
